bioneuron_oracle/builder.py

```diff
@@ -29,6 +31,22 @@ def __init__(self):
         self.spikes = neuron.h.Vector()
         self.ap_counter.record(neuron.h.ref(self.spikes))
         self.num_spikes_last = 0
+        self._clean = False
+
+    def update(self):
+        if self._clean:
+            raise RuntimeError("cannot update() after cleanup()")
+        count = len(self.spikes) - self.num_spikes_last
+        self.num_spikes_last = len(self.spikes)
+        return count, np.asarray(self.v_record)[-1]
+
+    def cleanup(self):
+        if self._clean:
+            raise RuntimeError("cleanup() may only be called once")
+        self.v_record.play_remove()
+        self.t_record.play_remove()
+        self.spikes.play_remove()
```

Comment:
More like the former. Does the simulator` __exit__()` get called at the end of `sim.run()` (general nengo question)? If so then I don't think either approach will leak memory...but I'm not sure how you would raise the `KeyboardInterrupt` in the middle of a `sim.run()` call :/